[Gradient Compression] Explicitly specify the dtype of the error tensor

wayi1 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #51094 [Gradient Compression] Refactor default_hooks.py and powerSGD_hook.py by creating a util function that make a vanilla allreduce future
* **#50985 [Gradient Compression] Explicitly specify the dtype of the error tensor**
* #50981 [Gradient Compression] Simplify the implementation of warm-start
* #50974 [Gradient Compression] Typo fixes in PowerSGD
* #50973 [Gradient Compression] Allow PowerSGD to run vallina allreduce for the first K iterations

Explicitly specify the dtype of error tensor when it is initialized by zeros.

Previously if the dtype of input tensor is FP16, the error tensor is still created in FP32, although later it will be assigned by another FP16 tensor (`input_tensor_cp` - `input_tensor`).

This change will make the dtype of error tensor look more clear.

Additionally, also explicitly specify the dtype if rank-1 tensor buffer is empty.

Original PR issue: Investigate Applying PowerSGD to Communication Hook for Gradient Compression #47202

Differential Revision: [D26034988](https://our.internmc.facebook.com/intern/diff/D26034988/)

Commit:
[Gradient Compression] Explicitly specify the dtype of the error tensor

Explicitly specify the dtype of error tensor when it is initialized by zeros.

Previously if the dtype of input tensor is FP16, the error tensor is still created in FP32, although later it will be assigned by another FP16 tensor (`input_tensor_cp` - `input_tensor`).

This change will make the dtype of error tensor look more clear.

Original PR issue: Investigate Applying PowerSGD to Communication Hook for Gradient Compression #47202

Differential Revision: [D26034988](https://our.internmc.facebook.com/intern/diff/D26034988/)

[ghstack-poisoned]

Commit:
Update on "[Gradient Compression] Explicitly specify the dtype of the error tensor"


Explicitly specify the dtype of error tensor when it is initialized by zeros.

Previously if the dtype of input tensor is FP16, the error tensor is still created in FP32, although later it will be assigned by another FP16 tensor (`input_tensor_cp` - `input_tensor`).

This change will make the dtype of error tensor look more clear.

Original PR issue: Investigate Applying PowerSGD to Communication Hook for Gradient Compression #47202

Differential Revision: [D26034988](https://our.internmc.facebook.com/intern/diff/D26034988/)

[ghstack-poisoned]

Commit:
Update on "[Gradient Compression] Explicitly specify the dtype of the error tensor"


Explicitly specify the dtype of error tensor when it is initialized by zeros.

Previously if the dtype of input tensor is FP16, the error tensor is still created in FP32, although later it will be assigned by another FP16 tensor (`input_tensor_cp` - `input_tensor`).

This change will make the dtype of error tensor look more clear.

Additionally, also explicitly specify the dtype if rank-1 tensor buffer is empty.

Original PR issue: Investigate Applying PowerSGD to Communication Hook for Gradient Compression #47202

Differential Revision: [D26034988](https://our.internmc.facebook.com/intern/diff/D26034988/)

[ghstack-poisoned]

rohan-varma commented on torch/distributed/algorithms/ddp_comm_hooks/powerSGD_hook.py:
```diff
@@ -551,8 +551,8 @@ def decompress(fut):
-            # memoize the local errors.
-            state.error_dict[bucket_index] = input_tensor_cp - input_tensor
+            # Memorize the local errors.
+            state.error_dict[bucket_index].copy_(input_tensor_cp - input_tensor)
```
Is there a reason for this change? IIUC a copy is more extensive, as `a = b` for tensors just makes them point to the same underlying storage, but this actually results in a full data copy?

SciPioneer replied:
Thanks! I thought it could save some peak memory footprint. Let me double confirm if there is a performance regression.

SciPioneer replied:
You are right. There is a performance loss due to `copy_`. Changed it back to assignment. Good catch!

Commit:
Update on "[Gradient Compression] Explicitly specify the dtype of the error tensor"


Explicitly specify the dtype of error tensor when it is initialized by zeros.

Previously if the dtype of input tensor is FP16, the error tensor is still created in FP32, although later it will be assigned by another FP16 tensor (`input_tensor_cp` - `input_tensor`).

This change will make the dtype of error tensor look more clear.

Additionally, also explicitly specify the dtype if rank-1 tensor buffer is empty.

Original PR issue: Investigate Applying PowerSGD to Communication Hook for Gradient Compression #47202

Differential Revision: [D26034988](https://our.internmc.facebook.com/intern/diff/D26034988/)

[ghstack-poisoned]

rohan-varma reviewed (approved):
LGTM now, thanks!

Commit:
Update on "[Gradient Compression] Explicitly specify the dtype of the error tensor"


Explicitly specify the dtype of error tensor when it is initialized by zeros.

Previously if the dtype of input tensor is FP16, the error tensor is still created in FP32, although later it will be assigned by another FP16 tensor (`input_tensor_cp` - `input_tensor`).

This change will make the dtype of error tensor look more clear.

Additionally, also explicitly specify the dtype if rank-1 tensor buffer is empty.

Original PR issue: Investigate Applying PowerSGD to Communication Hook for Gradient Compression #47202

Differential Revision: [D26034988](https://our.internmc.facebook.com/intern/diff/D26034988/)

[ghstack-poisoned]
